### ISSUE TITLE
Division by zero error on checkout if discount + $0 product in cart

### ISSUE THREAD

**Issue by mintplugins** [easydigitaldownloads/easy-digital-downloads#6901]
### Expected behavior
When I have a $0 product in the cart with a discount code, I expect the price to be $0. 

### Actual behavior
The page fails with a warning
`Warning: Division by zero in /app/public/wp-content/plugins/easy-digital-downloads/includes/cart/class-edd-cart.php on line 688`

### Steps to reproduce the behavior
1) Add a $0 product to the cart
2) Add a non-free product to the cart
3) Add a discount code to the cart
4) Remove the non-free product from the cart
5) You should see there error here

### Information (if a specific version is affected):
PHP Version: 7.2

EDD Version (or branch): master

WordPress Version: 5.0-alpha-43556

**Comment by JJJ** [comment]
The error is in the `get_item_discount_amount()` method, related to this:

```
$subtotal_percent  = ( ( $price * $item['quantity'] ) / $items_subtotal );
```

# Suggest 2 fixes:

## Fix the math, to handle free (zero price) products

```
$subtotal_quantity = ( $price * $item['quantity'] )
$subtotal_percent  = ! empty( $subtotal_quantity ) ? ( $subtotal_quantity / $items_subtotal ) : 0;
```

## Avoid applying discounts if `$price` is zero

Change
```
if ( ! empty( $discounts ) ) {
```
To:
```
if ( ! empty( $price ) && ! empty( $discounts ) ) {
```

Fix 1 addresses the faulty math, which is good to correct anyways.
Fix 2 addresses the logic that a free product could even be discounted, which is impossible.

**Comment by SDavisMedia** [comment]
I can't replicate this to save my life. I even updated WP to `5.0-alpha-43623` and converted to PHP 7.2.

When I follow the replication steps, I end up with a cart that has a free product but still has a discount: https://cl.ly/f9d0ad513ac2

Purchase goes through just fine.

**Comment by cklosowski** [comment]
@SDavisMedia it isn't a fatal error:
` Warning: Division by zero in /app/public/wp-content/plugins/easy-digital-downloads/includes/cart/class-edd-cart.php on line 688` so the purchase may still go through, can you veirfy that didn't show up in your logs somewhere?

**Comment by SDavisMedia** [comment]
Replicated. This requires a flat rate discount.

**Comment by SDavisMedia** [comment]
To me, it only makes sense to go with @jjj's second suggestion. There's no reason to run the logic over free cart items. And I don't see any hooks that'll allow the price of an item to change later in the process.

Only thing I wonder about is the fact that in the scenario outlined in the replication steps, you can complete the purchase of a free item with a discount applied to the cart. Obviously, the discount does nothing. _Should we_ be removing it from the cart?

**Comment by SDavisMedia** [comment]
> I think we leave them as is, even though it’s kinda weird?

I agree.

> I also think it might be a good idea to add a notice to the checkout page when the code does not actually changing the final dollar amount. If a customer is trying to add a code, I think it’s a safe assumption they expect the amount to change. If the amount doesn’t change (free items, or the code specifically omits the items being bought) it might reduce some support burden to say “We’ve added that discount code, but it didn’t reduce your total.”

@JJJ It may not be worth it to do that. Reason being, I can't think of a scenario where a customer can add a discount to the cart that doesn't apply to anything in the cart. 

* If the cart total is 0, no discount field is shown.
* If the discount field is shown and a discount is entered that does not apply to any items in the cart, an error is shown and the discount is not added.

So unless I'm forgetting a case, the only way to get a discount in the cart when it can't apply to anything in the cart is to perform the trickery outlined in the replication steps of this issue. Meaning when the discount is actually added, it did what it was supposed to do.

In other words, the scenario in this issue is probably a super edge case anyway.

**Comment by cklosowski** [comment]
Came here to agree, that discount codes need to stick around, even if they don't reduce the price of anything. I know of a few cases where they are used to offer a free follow up or additional service if a discount code is used, and for campaign tracking as well.

**Comment by SDavisMedia** [comment]
@cklosowski My PR should be all that's needed then.
